### PR TITLE
Add an example iOS app to prove out crates running on iOS.

### LINKED ISSUE #237: Swift bindings only work in tests
Specifically we include the `import uniffi_{{ ci.namespace() }}` statement at the beginning of these bindings, which I think is there because in tests we compile the Swift code into `libuniffi_{namespace}.dylib` to be able to be used into the Swift REPL. 
In regular production contexts the headers are included in the project and this statement causes a hard error.
@linacambridge I saw you worked on this, does what I describe makes sense?

┆Issue is synchronized with this [Jira Task](https://jira.mozilla.com/browse/SYNC-1655)

### PR DESCRIPTION
**Pull request by jhugman**
Fixes #237.

This PR looks larger than it really is.

The bulk of the file changes are in generating an iOS app from a standard SwiftUI template from Xcode.

Added: an `is_testing` argument passed down to the bindings, so the user doesn't have to think about the difference between `cargo test` and invoking `uniffi-bindgen` in different ways.
Modified: it's not clear to me how module maps work. However, the ones we were generating would use an absolute path for the header. This has changed to use a relative path.
Added: a `module_name` config variable, similar to `package_name` for kotlin package_name.

### CODE REVIEW

File: examples/app/ios/IOSApp/ContentView.swift [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+            }
+        }
+    }
+}
+
+extension TodoEntry: Hashable {
+    public func hash(into hasher: inout Hasher) {
+        hasher.combine(text)
+    }
+}
```
Comment (jhugman) [inline examples/app/ios/IOSApp/ContentView.swift:48]: We may want consider generating this for all our records.

File: examples/app/ios/IOSApp/ContentView.swift [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+struct ContentView_Previews: PreviewProvider {
+    static var previews: some View {
+        ContentView(todoList: TodoList())
+    }
```
Comment (jhugman) [inline examples/app/ios/IOSApp/ContentView.swift:53]: Wanted: `TodoListProtocol` that foreign languages can mock. See #315 .

File: examples/app/shared/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,19 @@
+[package]
+name = "uniffi-example-app"
+edition = "2018"
+version = "0.1.0"
+authors = ["Firefox Sync Team <sync-team@mozilla.com>"]
+license = "MPL-2.0"
+publish = false
+
+[lib]
+crate-type = ["staticlib", "cdylib"]
```
Comment (jhugman) [inline examples/app/shared/Cargo.toml:10]: We use `staticlib` for iOS and `cdylib` for testing.

File: uniffi_bindgen/src/bindings/swift/templates/wrapper.swift [hunk 4]
```diff
@@ -2,7 +2,9 @@
 // Trust me, you don't want to mess with it!
 
 import Foundation
-import uniffi_{{ ci.namespace() }}
+{% if is_testing -%}
+import {{ config.module_name() }}
+{% endif -%}
```
Comment (jhugman) [inline uniffi_bindgen/src/bindings/swift/templates/wrapper.swift:7]: This is the real meat of making it work.

The rest is convincing ourselves that it does.

File: examples/app/ios/IOSApp-Bridging-Header.h [hunk 5]
```diff
@@ -0,0 +1,11 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+
+#ifndef IOSApp_Bridging_Header_h
+#define IOSApp_Bridging_Header_h
+
+#import "todolist-Bridging-Header.h"
```
Comment (eoger) [inline examples/app/ios/IOSApp-Bridging-Header.h:9]: That reminds me we should really do https://github.com/mozilla/uniffi-rs/issues/266.

File: examples/app/ios/IOSApp.xcodeproj/project.xcworkspace/xcuserdata/jhugman.xcuserdatad/IDEFindNavigatorScopes.plist [hunk 6]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (eoger) [inline examples/app/ios/IOSApp.xcodeproj/project.xcworkspace/xcuserdata/jhugman.xcuserdatad/IDEFindNavigatorScopes.plist:1]: You might want to delete these `userdata` folders from your commit.

File: examples/app/shared/tests/bindings/test_todolist.kts [hunk 7]
```diff
@@ -0,0 +1,55 @@
+import uniffi.todolist.*
```
Comment (eoger) [inline examples/app/shared/tests/bindings/test_todolist.kts:1]: Do we want to copy the tests too?

File: uniffi_bindgen/src/bindings/swift/mod.rs [hunk 8]
```diff
@@ -31,26 +31,32 @@ pub fn write_bindings(
...
     let mut h = File::create(&header_file).context("Failed to create .h file for bindings")?;
     write!(h, "{}", header)?;
 
     let mut m =
         File::create(&module_map_file).context("Failed to create .modulemap file for bindings")?;
-    write!(m, "{}", generate_module_map(&ci, &header_file)?)?;
+    write!(
+        m,
+        "{}",
+        generate_module_map(config, &ci, &Path::new(&header_filename))?
```
Comment (eoger) [inline uniffi_bindgen/src/bindings/swift/mod.rs:58]: I wonder if we need the modulemap also? I don't think we do, so maybe we could in a future patch guard it with `is_testing`?

**Review by eoger — COMMENTED** [review]
Nice! Do you think it might be worth adding a README to the app dir explainings how we tied everything together in the XCode project? ("No" is a valid answer though, the uniffi goal is not to explain the full rust+ios integration steps).

File: examples/todolist/Cargo.toml [hunk 9]
```diff
@@ -7,7 +7,7 @@ license = "MPL-2.0"
 publish = false
 
 [lib]
-crate-type = ["cdylib"]
+crate-type = ["staticlib", "cdylib"]
```
Comment (jhugman) [inline examples/todolist/Cargo.toml:10]: Generate the `.a` file.

**Review by jhugman — COMMENTED** [review]
Addressed review comments; removed the second copy of todolist completely. This makes everything a lot tighter. 

There is scope for a README.md in `example/app` or `example/app/ios`, which should cover the targets to install, and the steps to getting it to run in Xcode. This might be a bit too bare bones, given that we'll almost certainly be going through a megazord or packaging via carthage.

Currently we can 

```sh
$ cargo clean 
$ cd example/app/ios
$ xcodebuild clean 
```

Then hitting play on Xcode.

**Review by eoger — APPROVED** [review]
The README is superb, thank you!

**Comment by eoger** [comment]
Note: we'll probably want to squash before merging.

**Pull request merged**